File: LeetCode/src/slidingwindow/MaximumAverageSubarrayI.java
package slidingwindow;

public class MaximumAverageSubarrayI {
    public double findMaxAverage(int[] nums, int k) {
        double average = Double.NEGATIVE_INFINITY, sum = 0;
        int n = nums.length;
        int left = 0, right = 0;

        while (right < n) {
            sum += nums[right];
            right++;

            if (right - left == k) {
                average = Math.max(average, sum / k);
                sum -= nums[left];
                left++;
            }
        }


        return average;
    }

    public static void main(String[] args) {
        MaximumAverageSubarrayI solution = new MaximumAverageSubarrayI();
        System.out.println(solution.findMaxAverage(new int[]{1, 12, -5, -6, 50, 3}, 4)); // Output: 12.75
        System.out.println(solution.findMaxAverage(new int[]{5}, 1)); // Output: 5
    }
}

File: LeetCode_Python/slidingwindow/MaximumAverageSubarrayI.py
class MaximumAverageSubarrayI:
    def findMaxAverage(self, nums: list[int], k: int) -> float:
        max_sum = sum(nums[:k])
        res = max_sum / k

        for i in range(k, len(nums)):
            max_sum += nums[i] - nums[i - k]
            res = max(res, max_sum / k)

        return res


# Time complexity: O(n), where n is the length of the nums list.
# Space complexity: O(1), as we are using a constant amount of space.

if __name__ == "__main__":
    solution = MaximumAverageSubarrayI()

    print("Expected Output: 12.75")
    print("Actual Output:", solution.findMaxAverage([1, 12, -5, -6, 50, 3], 4))

    print("Expected Output: 5.0")
    print("Actual Output:", solution.findMaxAverage([5], 1))
